Prevent cascade deletion of excluded tables #8

shuto-facengineer opened the pull request:
Hello! 
This PR addresses Issue #8 related to cascade deletion on interleaved tables.  

# Related Issue
- fixes #8 - Cascade deletion deletes child tables that must be excluded.

# Changes
- Updated `fetchTableSchemas` to change its execution mode based on the arguments provided.
- Updated `fetchTableSchemas` to exclude parent tables of specified exclude tables.
- Updated the README to explain the handling of interleaved tables.

In this PR, I have also introduced switch statement to avoid deeply nested if statements. Please review and consider whether these changes are acceptable to you.

shuto-facengineer committed:
Prevent cascade deletion of excluded tables

shuto-facengineer commented on truncate/table_schema.go:
```diff
@@ -96,18 +111,6 @@ func fetchTableSchemas(ctx context.Context, client *spanner.Client, targetTables
 			return err
 		}
 
-		if !truncateAll {
-			if len(excludes) != 0 {
-				if _, ok := excludes[tableName]; ok {
-					return nil
-				}
-			} else {
-				if _, ok := targets[tableName]; !ok {
-					return nil
-				}
-			}
-		}
-
```
To avoid excessive nesting from adding the new if statement, I have opted to use a switch statement instead.

shuto-facengineer commented on truncate/table_schema.go:
```diff
@@ -123,6 +126,20 @@ func fetchTableSchemas(ctx context.Context, client *spanner.Client, targetTables
 			}
 		}
 
+		switch mode {
+		case fetchModeTarget:
+			if _, ok := targets[tableName]; !ok {
+				return nil
+			}
+		case fetchModeExclude:
+			if _, ok := excludes[tableName]; ok {
+				if typ == deleteActionCascadeDelete {
+					excludeParents[parentTableName] = true
+				}
+				return nil
+			}
+		}
```
This switch statement handles processing based on the execution mode. fetchModeExclude records the parent table of exclude table that will be deleted by cascade deletion.

shuto-facengineer commented on truncate/table_schema.go:
```diff
@@ -134,6 +151,17 @@ func fetchTableSchemas(ctx context.Context, client *spanner.Client, targetTables
 		return nil, err
 	}
 
+	// Remove tables that may delete the exclude tables in cascade
+	if mode == fetchModeExclude && len(excludeParents) > 0 {
+		filtered := make([]*tableSchema, 0, len(tables))
+		for _, t := range tables {
+			if _, ok := excludeParents[t.tableName]; !ok {
+				filtered = append(filtered, t)
+			}
+		}
+		tables = filtered
+	}
```
When in fetchModeExclude, it becomes necessary to exclude the recorded parent tables.

yfuruyama commented:
@shuto-facengineer  LGTM. Thank you for fixing the long-standing bug!

If possible, can we add a unit test for `fetchTableSchemas`? Current signature has `client *spanner.Client` and it could make it difficult to write a unit test. So maybe we can divide current logic into 1) fetching the table list from Spanner, and 2) filtering out the tables to write a unit test easily.

shuto-facengineer commented:
@yfuruyama 
Thank for your review!
The changes might be a bit large for a single PR, but I'll attempt to split the logic and add unit tests.

shuto-facengineer committed:
Divide fetchTablesSchemas into fetching and filtering

shuto-facengineer committed:
Added unit test of filterTableSchemas

shuto-facengineer commented on truncate/index_schema.go:
```diff
@@ -0,0 +1,69 @@
+	"context"
+)
+
+// indexSchema represents secondary index metadata.
+type indexSchema struct {
+	indexName string
+
+	// Table name on which the index is defined.
+	baseTableName string
+
+	// Table name the index interleaved in. If blank, the index is a global index.
+	parentTableName string
+}
```
Separated code for indexSchema from tableSchema for readability.

yfuruyama replied:
I understand the purpose, but I don't think readability will be much improved by separating the file for index schema. Maybe putting all the code accessing `INFORMATION_SCHEMA` into a single file would be more readable in some situations.

shuto-facengineer replied:
> putting all the code accessing INFORMATION_SCHEMA into a single file would be more readable

This explanation made sense to me. I'll put them into a single file.

shuto-facengineer commented on truncate/table_schema.go:
```diff
@@ -43,45 +42,38 @@ type tableSchema struct {
+
+// fetchTableSchemas fetches schema information from spanner database.
+// If targetTables is not empty, it fetches only the specified tables.
+func fetchTableSchemas(ctx context.Context, client *spanner.Client, targetTables []string) ([]*tableSchema, error) {
```
fetchTableSchemas has been changed to concentrate on fetching schema information from Spanner.

yfuruyama replied:
Sounds good!

If `fetchTableSchemas` focuses on fetching schema from Spanner, how about moving `targetTables` filtering to `filterTableSchemas`?

So that each function will be:

- `fetchTableSchemas`: Fetches the schema information from Spanner
- `filterTableSchemas`: Filters out the unnecessary tables with `targetTables` and `excludedTables`

shuto-facengineer replied:
Ah, now I understand your intention! 
I will make the necessary modifications and request your review again in a few days.

shuto-facengineer commented on truncate/table_schema.go:
```diff
@@ -137,38 +123,46 @@ func fetchTableSchemas(ctx context.Context, client *spanner.Client, targetTables
-		SELECT INDEX_NAME, TABLE_NAME, PARENT_TABLE_NAME FROM INFORMATION_SCHEMA.INDEXES
-		WHERE INDEX_TYPE = 'INDEX' AND TABLE_CATALOG = '' AND TABLE_SCHEMA = '';
-	`))
+// filterTableSchemas filters tables by excludeTables.
+// If an exclude table is cascade deletable, its parent table is also excluded.
+func filterTableSchemas(tables []*tableSchema, excludeTables []string) []*tableSchema {
```
New function to filter fetched tableSchema by excludeTables.

yfuruyama replied:
Thanks!

shuto-facengineer commented on truncate/table_schema.go:
```diff
@@ -137,38 +123,46 @@ func fetchTableSchemas(ctx context.Context, client *spanner.Client, targetTables
+				}
+			}
 		}
 
-		var parentTableName string
-		if parent.Valid {
-			parentTableName = parent.StringVal
+		// loop until no more parent tables to exclude
+		if len(excludeParents) == 0 {
+			break
 		}
 
-		indexes = append(indexes, &indexSchema{
-			indexName:       indexName,
-			baseTableName:   baseTableName,
-			parentTableName: parentTableName,
-		})
-		return nil
-	}); err != nil {
-		return nil, err
+		for p := range excludeParents {
+			excludes[p] = struct{}{}
+		}
+	}
```
This part might be a little less readable. 
Previous commits did not support hierarchical interleaving. Therefore, I have added a process that traces back to the top level hierarchy.

yfuruyama replied:
Yeah, let me review this once another part is changed.

shuto-facengineer commented on truncate/table_schema_test.go:
```diff
@@ -0,0 +1,121 @@
+		excludeTables []string
+		want          []*tableSchema
+	}{
+		"Exclude the parent tables by tracing up to the topmost level.": {
+			schemas:       []*tableSchema{singers, albums, songs, t1, t2, t3},
+			excludeTables: []string{songs.tableName},
+			want:          []*tableSchema{t1, t2, t3},
+		},
+		"Exclude only the higher levels without the lower levels.": {
+			schemas:       []*tableSchema{singers, albums, songs, t1, t2, t3},
+			excludeTables: []string{albums.tableName},
+			want:          []*tableSchema{songs, t1, t2, t3},
+		},
+		"Exclude multiple tables.": {
+			schemas:       []*tableSchema{singers, albums, songs, t1, t2, t3},
+			excludeTables: []string{songs.tableName, t1.tableName, t2.tableName},
+			want:          []*tableSchema{t3},
+		},
+		"Do nothing when no exclude tables are passed.": {
+			schemas:       []*tableSchema{singers, albums, songs, t1, t2, t3},
+			excludeTables: nil,
+			want:          []*tableSchema{singers, albums, songs, t1, t2, t3},
+		},
```
Added simple unit tests.

shuto-facengineer committed:
Applied format

shuto-facengineer commented:
@yfuruyama I have modified the code based on your suggestion ! 
This task was good to learn coding for me. Thank you !

# Changes
- Divided fetchTableSchemas into two parts: fetch and filter.
- Added a process that traces back to the top level hierarchy in order to exclude all ancestor tables.
- Added simple unit tests of filter logic.
- Separated the code for indexSchema from tableSchema ( to a separated file ).

I did not get around to implementing unit tests for fetchTableSchemas because I thought the scope of this Pull Request should be limited to bug fixes for exclude.

If the volume of changes is too large to review, I will try to separate the PRs in stages. Please feel free to let me know.

yfuruyama commented on truncate/table_schema.go:
```diff
@@ -43,45 +43,38 @@ type tableSchema struct {
 }
 
-func fetchTableSchemas(ctx context.Context, client *spanner.Client, targetTables, excludeTables []string) ([]*tableSchema, error) {
+const fetchTableSchemasQuery = `
```
Is there any reason this query is defined as constant?

This is the personal opinion, but I would prefer having this query in the `client.Single().Query()` code because it doesn't force the reader to move their eyes back and forth.

shuto-facengineer replied:
This change was more a quirk of my own coding habits than a deliberate design choice.
I'll revert the change. Thank you!

yfuruyama commented on truncate/table_schema_test.go:
```diff
@@ -0,0 +1,120 @@
+)
+
+func TestFilterTableSchemas(t *testing.T) {
+	t.Parallel()
```
`t.Parallel()` might not be needed as each test is light-weight.

yfuruyama commented on truncate/table_schema_test.go:
```diff
@@ -0,0 +1,120 @@
+		excludeTables []string
+		want          []*tableSchema
+	}{
+		"Exclude the parent tables by tracing up to the topmost level.": {
```
To have a consistent style with existing tests, can we use `desc` (description) field to describe a test?

ref: https://github.com/cloudspannerecosystem/spanner-truncate/blob/main/truncate/coordinator_test.go#L34

shuto-facengineer replied:
Sure! I'll rewrite it in the same format.

yfuruyama commented on truncate/table_schema_test.go:
```diff
@@ -0,0 +1,120 @@
+
+			got := filterTableSchemas(tc.schemas, tc.excludeTables)
+
+			if len(got) != len(tc.want) {
+				t.Errorf("len(got) %d, len(want) %d", len(got), len(tc.want))
+			}
+
+			sort.Slice(got, func(i, j int) bool {
+				return got[i].tableName < got[j].tableName
+			})
+			sort.Slice(tc.want, func(i, j int) bool {
+				return tc.want[i].tableName < tc.want[j].tableName
+			})
+
+			for i, w := range tc.want {
+				if !reflect.DeepEqual(*got[i], *w) {
+					t.Errorf("got %v, want %v", got[i], w)
+				}
+			}
```
We can use [cmp](https://pkg.go.dev/github.com/google/go-cmp/cmp) package to compare the slices.

Also there is an option called [SortSlices](https://pkg.go.dev/github.com/google/go-cmp@v0.6.0/cmp/cmpopts#SortSlices) that can be used to sort slices before comparing them.

I found a good example here: https://zenn.dev/tminamiii/articles/go-cmp-in-unit-test#sortxxxx

yfuruyama commented on truncate/table_schema_test.go:
```diff
@@ -0,0 +1,120 @@
+}
+
+// The table schemas are well known in Cloud Spanner document about 'schema and data model'.
+var (
```
I think we can define these variables in `TestFilterTableSchemas` function to narrow the scope of variables.

yfuruyama reviewed (commented):
@shuto-facengineer Thank you so much for additional work to implement the unit tests. I really appreciate it.

I have commented some parts, so please take a look when you have a chance. Thank you!

shuto-facengineer committed:
Put index_schema.go into table_schema.go

shuto-facengineer committed:
Directly pass a query for fetching table schema

shuto-facengineer committed:
Fix TestFilterTableSchemas

shuto-facengineer committed:
Move target logic from fetch to filter

shuto-facengineer committed:
Revert unnecessary change.

shuto-facengineer committed:
Improve filterTableSchemas switch statement

shuto-facengineer commented on truncate/table_schema.go:
```diff
@@ -137,6 +118,93 @@ func fetchTableSchemas(ctx context.Context, client *spanner.Client, targetTables
 	return tables, nil
 }
 
+// filterTableSchemas filters tables with given targetTables and excludeTables.
+// If targetTables is not empty, it fetches only the specified tables.
+// If excludeTables is not empty, it excludes the specified tables.
+// TargetTables and excludeTables cannot be specified at the same time.
+func filterTableSchemas(tables []*tableSchema, targetTables, excludeTables []string) ([]*tableSchema, error) {
+	isExclude := len(excludeTables) > 0
+	isTarget := len(targetTables) > 0
+
+	switch {
+	case isTarget && isExclude:
+		return nil, errors.New("both targetTables and excludeTables cannot be specified at the same time")
+	case isTarget && !isExclude:
+		return targetFilterTableSchemas(tables, targetTables), nil
+	case !isTarget && isExclude:
+		return excludeFilterTableSchemas(tables, excludeTables), nil
+	default: // No target and exclude tables are specified.
+		return tables, nil
+	}
+}
```
The filterTableSchemas function calls different functions based on the arguments used for filtering.

shuto-facengineer commented on truncate/table_schema_test.go:
```diff
@@ -0,0 +1,225 @@
+		}
+	})
+
+	// Detailed tests for targetFilterTableSchemas
+	t.Run("targetFilterTableSchemas", func(t *testing.T) {
+		for _, test := range []struct {
+			desc         string
+			schemas      []*tableSchema
+			targetTables []string
+			want         []*tableSchema
+		}{
+			{
+				desc:         "Include multiple tables",
+				schemas:      []*tableSchema{t1, t2, t3},
+				targetTables: []string{t1.tableName, t2.tableName},
+				want:         []*tableSchema{t1, t2},
+			},
+			{
+				desc:         "Do nothing when no target tables are passed.",
+				schemas:      []*tableSchema{t1, t2, t3},
+				targetTables: nil,
+				want:         []*tableSchema{t1, t2, t3},
+			},
+			// TODO: Determine the specifications for parent-child relationships in hierarchical interleaved tables, and add corresponding tests and implementation.
+			// This includes defining the behavior of targetFilterTableSchemas for cases where tables not included in the target list are subject to cascade deletion.
+		} {
```
Currently, when a parent table is specified as a target table, interleaved child tables are not printed. However, I am considering changing the specification to print these interleaved child tables as well. What are your thoughts on this?

yfuruyama replied:
Sounds good. If I understand correctly, the final outcome won't be changed (i.e. interleaved tables are also deleted) and it just changes how the target tables are prompted to users?

If so, maybe different PR might be better 👍

shuto-facengineer replied:
Your understanding is correct.
I've opened a new issue regarding this problem (#18).

I noticed that Spanner already supports cascade deletion for foreign keys (see [Release Notes, July 18, 2023](https://cloud.google.com/spanner/docs/release-notes?hl=ja#July_18_2023)), so I've opened another issue (#19) as well.

In this PR, we focus on the #8 for cascade deletion on interleaved tables.

shuto-facengineer commented:
@yfuruyama 
I've implemented the changes based on your feedback. Please take a look and let me know if there are any further adjustments needed. Thank you!

# Changes
- Deleted `index_shema.go` and put the code into `table_schema.go`
- Deleted target filtering from `fetchTableSchemas`
- Added new functions ( targetFilterTableSchemas and excludeFilterTableSchemas)
- Added unit tests for filterTableSchemas, targetFilterTableSchemas and excludeFilterTableSchemas

# Discussion
- Similar to the consideration in the exclude logic, should the target logic also take into account the cascade deletion of child tables?

yfuruyama commented on truncate/table_schema.go:
```diff
@@ -137,6 +118,93 @@ func fetchTableSchemas(ctx context.Context, client *spanner.Client, targetTables
+	switch {
+	case isTarget && isExclude:
+		return nil, errors.New("both targetTables and excludeTables cannot be specified at the same time")
+	case isTarget && !isExclude:
```
Maybe we don't need `&& !isExclude`?

yfuruyama commented on truncate/table_schema.go:
```diff
@@ -137,6 +118,93 @@ func fetchTableSchemas(ctx context.Context, client *spanner.Client, targetTables
+		return nil, errors.New("both targetTables and excludeTables cannot be specified at the same time")
+	case isTarget && !isExclude:
+		return targetFilterTableSchemas(tables, targetTables), nil
+	case !isTarget && isExclude:
```
Same here?

yfuruyama commented on truncate/table_schema.go:
```diff
@@ -137,6 +118,93 @@ func fetchTableSchemas(ctx context.Context, client *spanner.Client, targetTables
+		return tables
+	}
+
+	targets := make(map[string]struct{}, len(targetTables))
```
Maybe `map[string]bool` would be simpler?

shuto-facengineer replied:
`struct{}` is a zero-size type, making it more memory-efficient than using `bool`. (found a good explanation: https://dave.cheney.net/tag/struct )
In this situation, it's clear that we are only interested in whether the keys exist, without implying any additional information. 

Originally, I considered using `map[string]struct{}` to be the appropriate implementation. However, using `if isTarget[key] { ... }` makes it immediately clear whether a key is present. So I decided to proceed with map[string]bool for its superior readability and ease of understanding.

```go
isTarget := make(map[string]bool, len(targetTables))
for _, t := range targetTables {
        isTarget[t] = true
}

filtered := make([]*tableSchema, 0, len(tables))
for _, t := range tables {
        if isTarget[t.tableName] { // No need to use 'if _, ok := isTarget[t.tableName]; ok {'
                filtered = append(filtered, t)
	}
}
```

yfuruyama commented on truncate/table_schema.go:
```diff
@@ -137,6 +118,93 @@ func fetchTableSchemas(ctx context.Context, client *spanner.Client, targetTables
+		return tables
+	}
+
+	excludes := make(map[string]struct{}, len(tables))
```
Maybe `map[string]bool` would be simpler?

yfuruyama commented on truncate/table_schema.go:
```diff
@@ -137,6 +118,93 @@ func fetchTableSchemas(ctx context.Context, client *spanner.Client, targetTables
+
+	// Add parent tables that may delete the exclude tables in cascade
+	// Since interleave tables can be hierarchical, tracing up to the top level is needed.
+	for {
+		excludeParents := make(map[string]struct{}, len(excludes))
+		for _, t := range tables {
+			if _, ok := excludes[t.tableName]; ok && t.isCascadeDeletable() {
+				if _, alreadyExcluded := excludes[t.parentTableName]; !alreadyExcluded {
+					excludeParents[t.parentTableName] = struct{}{}
+				}
+			}
+		}
+
+		// loop until no more parent tables to exclude
+		if len(excludeParents) == 0 {
+			break
+		}
+
+		for p := range excludeParents {
+			excludes[p] = struct{}{}
+		}
+	}
```
hmm... this would a bit hard to comprehend how it works.

How about splitting the entire logic into two steps - 1) find the all excluded tables recursively, and then 2) exclude tables?

In pseudo code:
```go
func findExcludeTables(tables []*tableSchema, excludeTables []string) []string {
	excludes := make(map[string]bool, len(tables))
	for _, t := range excludeTables {
		excludes[t] = true
	}

	var allExcludes []string
	for _, t := range tables {
		if _, ok := excludes[t.tableName]; ok {
			allExcludes = append(allExcludes, t.tableName)
			if t.isCascadeDeletable() {
				excluded := findExcludeTables(tables, []string{t.parentTableName})
				allExcludes = append(allExcludes, excluded...)
			}
		}
	}
	return allExcludes
}

func excludeFilterTableSchemas(tables []*tableSchema, excludeTables []string) []*tableSchema {
	excludes := findExcludeTables(tables, excludeTables)
	...
}
```

yfuruyama commented on truncate/table_schema_test.go:
```diff
@@ -0,0 +1,225 @@
+	"github.com/google/go-cmp/cmp/cmpopts"
+)
+
+func TestFilterTableSchemas(t *testing.T) {
```
In Go, it's common to have different test func for different package function, so maybe it'd better to have `TestFilterTableSchemas`, `TestTargetFilterTableSchemas`, and `TestExcludeFilterTableSchemas`.

The `filterTableSchemas` function is very straightforward, so I think it's ok to skip wrighting`TestFilterTableSchemas`.

yfuruyama reviewed (commented):
@shuto-facengineer Thank you again for your diligent effort on this change! I have left some comments, so when you have time, please take a look.

shuto-facengineer committed:
Change map[string]struct{} into map[string]bool.

shuto-facengineer committed:
Separated TestFilterTableSchemas into TestTargetFilterTableSchemas and TestExcludeFilterTableSchemas.

shuto-facengineer committed:
Improve switch case statement

shuto-facengineer committed:
Refactored the filter logic to use recursion.

shuto-facengineer committed:
Applied format

shuto-facengineer commented on truncate/table_schema.go:
```diff
@@ -54,7 +63,17 @@ type indexSchema struct {
 }
 
-func fetchTableSchemas(ctx context.Context, client *spanner.Client, targetTables, excludeTables []string) ([]*tableSchema, error) {
+// tableLineage represents a table schema and its ancestors and descendants.
+// This is used to represent inter-table relationships.
+// The order of ancestors and descendants are not guaranteed.
+type tableLineage struct {
+	tableSchema *tableSchema
+	ancestors   []*tableSchema
+	descendants []*tableSchema
+}
```
Introduced a new struct to encapsulate the logic for managing table relationships, enhancing code readability and maintainability.

And we can use descendants to fix #18 .

shuto-facengineer commented on truncate/table_schema.go:
```diff
@@ -137,6 +131,136 @@ func fetchTableSchemas(ctx context.Context, client *spanner.Client, targetTables
+	return filtered
+}
+
+// excludeFilterTableSchemas filters tables with given excludeTables.
+// If excludeTables is empty, it returns all tables.
+// When an exclude table is cascade deletable, its parent table is also excluded.
+func excludeFilterTableSchemas(tables []*tableSchema, excludeTableSchemas []string) []*tableSchema {
+	if len(excludeTableSchemas) == 0 {
+		return tables
+	}
+
+	isExclude := make(map[string]bool, len(tables))
+	for _, t := range excludeTableSchemas {
+		isExclude[t] = true
+	}
+
+	// Additionally exclude parent tables that may delete the exclude tables in cascade
+	lineages := constructTableLineages(tables)
+	for _, l := range lineages {
+		if isExclude[l.tableSchema.tableName] {
+			for _, a := range l.ancestors {
+				isExclude[a.tableName] = true
+			}
+		}
+	}
+
+	filtered := make([]*tableSchema, 0, len(tables))
+	for _, t := range tables {
+		if !isExclude[t.tableName] {
+			filtered = append(filtered, t)
+		}
+	}
+
+	return filtered
+}
```
Extracted ancestor and descendant retrieval into an another function, simplifying the code.

shuto-facengineer commented on truncate/table_schema.go:
```diff
@@ -137,6 +131,136 @@ func fetchTableSchemas(ctx context.Context, client *spanner.Client, targetTables
+	return lineages
+}
+
+// findAncestors recursively finds all ancestors of the given table .
+func findAncestors(table *tableSchema, parentRelation map[string]*tableSchema, ancestors []*tableSchema) []*tableSchema {
+	if !table.isRoot() {
+		parent := parentRelation[table.tableName]
+		ancestors = append(ancestors, parent)
+		return findAncestors(parent, parentRelation, ancestors)
+	}
+	return ancestors
+}
+
+// findDescendants recursively finds all descendants of the given table.
+func findDescendants(table *tableSchema, childRelation map[string][]*tableSchema, descendants []*tableSchema) []*tableSchema {
+	for _, child := range childRelation[table.tableName] {
+		descendants = append(descendants, child)
+		descendants = findDescendants(child, childRelation, descendants)
+	}
+	return descendants
+}
```
Implemented hierarchical traversal using recursion, resulting in relatively readable code. Thank you for your advice!

shuto-facengineer commented:
@yfuruyama 
Thank you for your review! I addressed your review comments! Please review the change when you have time. 

# Changes
- Introduced a new struct `tableLineage` to enhancing code readability and maintainability. This struct encapsulate concerns related to interleave relations.
- Implemented `constructTableLineages`. This function calls `findAncestors` and `findDescendants` to retrieve interleaved hierarchical relations.
- Separeted the test into `TestTargetFilterTableSchemas` and `TestExcludeFilterSchemas`.
- Adressed other review remarks.

shuto-facengineer committed:
Fix the bug building parentRelation and childRelation.

shuto-facengineer committed:
Optimized redundant for statement.

shuto-facengineer committed:
Eliminated the subtest structure.

shuto-facengineer commented on truncate/table_schema.go:
```diff
@@ -137,6 +131,130 @@ func fetchTableSchemas(ctx context.Context, client *spanner.Client, targetTables
+	return lineages
+}
+
+// findAncestors recursively finds all ancestors of the given table .
+func findAncestors(table *tableSchema, parentRelation map[string]*tableSchema, ancestors []*tableSchema) []*tableSchema {
+	if parent, ok := parentRelation[table.tableName]; ok {
+		ancestors = append(ancestors, parent)
+		return findAncestors(parent, parentRelation, ancestors)
+	}
+	return ancestors
+}
+
+// findDescendants recursively finds all descendants of the given table.
+func findDescendants(table *tableSchema, childRelation map[string][]*tableSchema, descendants []*tableSchema) []*tableSchema {
+	for _, child := range childRelation[table.tableName] {
+		descendants = append(descendants, child)
+		descendants = findDescendants(child, childRelation, descendants)
+	}
+	return descendants
+}
```
Implemented hierarchical traversal using recursion, resulting in relatively readable code. Thank you for your advice!

shuto-facengineer commented:
@yfuruyama 
Thank you for your review! I addressed your review comments! Please review the change when you have time. 

# Changes
- Introduced a new struct `tableLineage` to enhancing code readability and maintainability. This struct encapsulate concerns related to interleave relations.
- Implemented `constructTableLineages`. This function calls `findAncestors` and `findDescendants` to retrieve interleaved hierarchical relations.
- Separeted the test into `TestTargetFilterTableSchemas` and `TestExcludeFilterSchemas`.
- Adressed other review remarks.

yfuruyama commented on truncate/table_schema.go:
```diff
@@ -137,6 +131,130 @@ func fetchTableSchemas(ctx context.Context, client *spanner.Client, targetTables
+	parentRelation := make(map[string]*tableSchema, len(tables))
+	childRelation := make(map[string][]*tableSchema, len(tables))
+	for _, t := range tables {
+		if !t.isRoot() && t.isCascadeDeletable() {
```
If we construct lineages for tables, it might not be appropriate to use `t.isCascadeDeletable()`. Whether or not it's cascading deleted, I think the child table should be semantically in the table lineages.

yfuruyama reviewed (approved):
Thank you for the changes! LGTM. I commented minor suggestion.

shuto-facengineer committed:
Remove cascade deletability check from constructTableLineages

shuto-facengineer committed:
Add test to ensure parents of tables with ON DELETE NO ACTION are not excluded

shuto-facengineer commented:
@yfuruyama 
Thank you for LGTM and the comment.
I addressed the comment(0346895a9b83dd0c5bff0875590dc86a279e8653) and added the test to ensure the entire logic is not changed(5564915b96cad1db5844bd033189d658fcf97dce).

After this PR is merged, I will create a PR for #18 as well.

yfuruyama commented:
@shuto-facengineer Thank you for your continuous effort on improving this PR! LGTM. I'll merge this.
